### ISSUE TITLE
[vLLM Plugin] Garbage results when using non-greedy SamplingParams (bitonic sort duplicate/missing indices bug)

### ISSUE THREAD

**Issue by kmabeeTT** [tenstorrent/tt-xla#3253]
**Details:**

With 2 fixes for #3027 fails now in place (runtime asserts from Sort op) we now see garbage outputs which as it turns out is also coming from Sort op.   Opened https://github.com/tenstorrent/tt-metal/issues/37571 to track proper fix.

**Repro:**

Remove `or True` hack in ` here in `model_runner.py` that forces greedy-only:

```
        # @LPanosTT: sampler functionalitty has issues currently, so we will always use greedy sampling
        if sampling_metadata.all_greedy or True:
```

Then run upcoming `tests/integrations/vllm_plugin/generative/test_sampling_params.py::test_temperature` test (FIXME - Cherry pick) and see:

```
The capital of France is

[TESTOUT test_temperature] temp=0.0:  the capital of the French Republic.
[TESTOUT test_temperature] temp=0.5:  named after theari named after the named afteract...
[TESTOUT test_temperature] temp=0.8: Integer namedInstoreAndOnline named after the)mad...
[TESTOUT test_temperature] temp=1.0:  the world capital resembled a attack on the Neptu...
[TESTOUT test_temperature] temp=1.5: /" Dock alienatedGAME character70710inki fabrics i...
```

There is a workaround that we can use in meantime, will reply here separetely...

**Comment by kmabeeTT** [comment]
For the garbage output, there is a one line workaround/fix after sampler creation we can use here to avoid the bug:

```
diff --git a/integrations/vllm_plugin/vllm_tt/model_runner.py b/integrations/vllm_plugin/vllm_tt/model_runner.py
index 6a120c5dc..9b9ed8e40 100644
--- a/integrations/vllm_plugin/vllm_tt/model_runner.py
+++ b/integrations/vllm_plugin/vllm_tt/model_runner.py
@@ -1354,6 +1354,14 @@ class TTModelRunner(LoRAModelRunnerMixin, KVConnectorModelRunnerMixin):
             self.model = model
         self.model.compile(backend="tt", dynamic=False)
         self.sampler = TPUSampler()
+
+        # Use TPU top-k/top-p path which thresholds logits directly
+        # instead of sort+scatter. This should be more performant
+        # and avoids ttnn bitonic sort bug.
+        # See https://github.com/tenstorrent/tt-xla/issues/3253
+        self.sampler.topk_topp_sampler.forward = (
+            self.sampler.topk_topp_sampler.forward_tpu
+        )
         logger.info(f"Compiled model: \n{self.model}")
```


Here's what it does:

```
  The VLLM TopKTopPSampler has 3 forward methods:
  - forward_native — selected automatically for TT (since TT is neither CUDA nor TPU)
  - forward_tpu — designed for TPU
  - forward_cuda — uses FlashInfer
```

The workaround overrides the method pointer from `forward_native` to `forward_tpu`. What's different between them?

`forward_native` does top-k/top-p filtering by **rearranging** the logits:
```
  logits_sort, logits_idx = logits.sort(...)     # sort
  logits_sort.masked_fill_(top_p_mask, -inf)     # filter in sorted order
  logits = logits_sort.scatter(..., logits_idx)  # put back in original positions ← BROKEN
```

 `forward_tpu` does top-k/top-p filtering by **thresholding** the original logits:
```
  probs_sort, _ = probs.sort(...)                  # sort (indices discarded with _)
  top_k_cutoff = probs_sort.gather(-1, count)      # find the cutoff value
  logits.masked_fill_(probs < top_k_cutoff, -inf)  # mask original logits directly
```

Both achieve the same result — zeroing out low-probability tokens — but `forward_tpu` never needs to scatter values back using sort indices. It just asks "what's the threshold probability?" and masks anything below it in-place on the original
   unsorted tensor.

On the question of whether this should be considered a hack/workaround or a more legitimate fix..

  1. **`forward_tpu` was specifically designed to avoid scatter** — the TPU team wrote it because scatter is slow on TPU. The comment in vllm's source says exactly this: "This algorithm avoids using torch.scatter which is extremely slow on TPU." TT
   has the same problem (scatter with sort indices is broken), so using the scatter-free path is the right architectural choice.
  2. **Both paths are functionally equivalent** — they implement the same top-k/top-p filtering, just with different algorithms. The TPU path isn't a degraded or approximate version.
  3. **TT should arguably have been using this path from the start** — the only reason it doesn't is that vllm's platform detection falls through to forward_native as the default for unknown platforms. TT is much more similar to TPU than to CUDA
  in terms of what operations are efficient.

**Comment by kmabeeTT** [comment]
Marking closed. Was going to submit that proposed workarond/fix, but vllm v0.13.0 uplift landed first which actually made that the default behavior for TPUSampler, so this got resolved via #3272.

The test_sampling_params_sweep.py was checked in as `tests/integrations/vllm_plugin/sampling/test_sampling_params.py::test_sampling_param_sweep[temperature]` just now at bd8598551 and gives non-garbage output now (different prompt, here is the previous prompt shown in this ticket's output for reference):

```
//Before Fix

The capital of France is
[TESTOUT test_temperature] temp=0.0:  the capital of the French Republic.
[TESTOUT test_temperature] temp=0.5:  named after theari named after the named afteract...
[TESTOUT test_temperature] temp=0.8: Integer namedInstoreAndOnline named after the)mad...
[TESTOUT test_temperature] temp=1.0:  the world capital resembled a attack on the Neptu...
[TESTOUT test_temperature] temp=1.5: /" Dock alienatedGAME character70710inki fabrics i...

// After Fix:

The capital of France is
[TESTOUT test_sampling_param_sweep] temperature=0.5:  still a country that has been ruled by a Communis...
[TESTOUT test_sampling_param_sweep] temperature=0.8:  not broken up, but the whole world is made up of ...
[TESTOUT test_sampling_param_sweep] temperature=1.0:  the most deprived area of Europe. It has an urban...
[TESTOUT test_sampling_param_sweep] temperature=1.5:  giving customers temporary gown doesn't allow the...
```
